demos/sjone/ir_receiver/source/main.cpp

```diff
@@ -0,0 +1,146 @@
+#include "L1_Peripheral/lpc17xx/system_controller.hpp"
+#include "L1_Peripheral/lpc17xx/pulse_capture.hpp"
+#include "L1_Peripheral/lpc17xx/timer.hpp"
+#include "L2_HAL/communication/tsop752.hpp"
+#include "utility/log.hpp"
+
+struct DecodedFrame_t
+{
+  uint32_t data;
+  bool is_valid;
+};
+
+constexpr DecodedFrame_t kInvalidFrame = {
+  .data     = 0,
+  .is_valid = false,
+};
+
+bool IsDurationWithinTolerance(uint32_t duration,
+                               std::chrono::microseconds expected_duration,
+                               float tolerance)
+{
+  uint32_t min = static_cast<uint32_t>(
+      static_cast<float>(expected_duration.count()) * (1.0f - tolerance));
+  uint32_t max = static_cast<uint32_t>(
+      static_cast<float>(expected_duration.count()) * (1.0f + tolerance));
+  return (min <= duration) && (duration <= max);
+}
+/// The Samsung BN59-00678AS remote uses SPACE_ENC (pulse distance encoding).
+/// Meaning the logic level can be determined by examining the duration of each
+/// data pulse's "space" (the duration when the pulse is low).
+///
+/// Example:
+///    _____________________                       ___________
+///   |                     |                     |
+///   ^                     v                     ^
+///   |                     |_____________________|
+///
+///   |<------ mark ------->|<------ space ------>|
+///
+/// @see: http://lirc.sourceforge.net/remotes/samsung/BN59-00678A
+const DecodedFrame_t DecodeSamsungRemote(

```

Comment:
This may be good to put this algorithm somewhere in the libraries.

Reply:
Would it be better to put this in the InfraredReceiver class as a helper function or to put it in a utility/infrared.hpp? This function can be modified a bit to be a generic decoding algorithm for pulse width encoding.

Reply:
I'm honestly not sure where the best place would be. I think it would make the most sense in the InfraredReceiver header. Not sure if having it within the class makes the most sense. Could just be an inline function. It could also be in another file named "infrared_algorithms.h". 

Making it generic would be great!